### PR TITLE
Support not passing metadata to socket-based services

### PR DESCRIPTION
**Pull request by DemiMarie**
This makes it easier to implement socket-based services that do not
require the metadata.  This avoids having to use a slow executable-based
service or write a custom wrapper.

Fixes: https://github.com/QubesOS/qubes-issues/issues/9036

Edit: This PR now also fixes https://github.com/QubesOS/qubes-issues/issues/9073.  Furthermore, the first four commits fix multiple bugs in the test suite, and the fifth and sixth are purely test suite enhancements.

### CODE REVIEW

**Comment by marmarek** [comment]
There are unit tests for socket services (look for `qubes.SocketService`), can you extend them to check this new option too?

**Comment by DemiMarie** [comment]
> There are unit tests for socket services (look for `qubes.SocketService`), can you extend them to check this new option too?

Added.  They caught a nasty bug in the TOML code as well.

File: qrexec/tests/socket/agent.py [hunk 1]
```diff
@@ -438,6 +438,41 @@ def test_exec_service_with_arg(self):
...
+        self.addCleanup(server.close)
+
+        target = self.execute_qubesrpc("qubes.SocketService+arg2", "domX")
+
+        server.accept()
+
+        message = b"stdin data"
+        target.send_message(qrexec.MSG_DATA_STDIN, message)
+        target.send_message(qrexec.MSG_DATA_STDIN, b"")
+        self.assertEqual(server.recvall(len(message) - 1), message[:-1])
```
Comment (marmarek) [inline qrexec/tests/socket/agent.py:461]: Why -1?

Comment (DemiMarie) [reply qrexec/tests/socket/agent.py]: I was trying to figure out why the test was hanging.  It turns out that it was failing to load the config file.

**Comment by marmarek** [comment]
PipelineRetry

**Comment by marmarek** [comment]
PipelineRetry

**Comment by DemiMarie** [comment]
> Finally, we have test coverage reporting here :)
> 
> I see it claims some parts of `load_service_config_v2()` are not covered, that I think do have some tests already (especially setting wait-for-session=1), can you verify it? (maybe it's another case where test-provided config isn't actually used? or maybe it's a legitimate code issue?)
> 
> And also, I see some low hanging fruits regarding toml parser:
> 
> * string option (add any test setting `user='user'`, even though actual user switching isn't supported in unit tests)
> * false boolean value, integer starting with 0

Right now, it’s very hard to test (or debug!) services that either invalid command lines or invalid configs.  The connection simply hangs with no error message.  I could add a hacky workaround, but I’d prefer a proper solution: have the agent send `MSG_SERVICE_REFUSED` in this case.  I believe @omeg will be working on this.

**Comment by marmarek** [comment]
> have the agent send `MSG_SERVICE_REFUSED` in this case

No, that wont work, as this is only daemon->agent message (for a reason). At this stage, the calling qube already tries to establish data vchan, and it needs to get a failure message somehow (or wait until timeout, which is less helpful). So, the proper error handling at this stage is to establish that data vchan and send error as an exit code. Sending also an error message on stderr might be an option, but I'd be careful about that, as it might be confusing.

**Comment by DemiMarie** [comment]
Now fixed in f7ec4b0f67f2f730f639dcda249310cae647294c.  Additionally, 518ec17757092a6be58f6d8c31a9789939db7995 adds support for running tests with sanitizers and 07e2c3c31fa8d118af606e64e06c0f791278e5cb fixes what I believe is a race condition in the test suite.  I had to use clang for ASAN+UBSAN builds, as binaries built with GCC crashed with SIGSEGV.

**Review by marmarek — APPROVED** [review]

**Comment by DemiMarie** [comment]
This has a bug: skip-service-descriptor is _not_ honored for calls to dom0.  #136 fixes this.

**Pull request merged**